wagtail/models/__init__.py

```diff
@@ -2910,13 +2911,103 @@ class UserPagePermissionsProxy:
+                "group_permissions",
+                queryset=GroupPagePermission.objects.filter(group__user=user),
+                to_attr="_perms_for_user",
+            )
+            # Load all the distinct pages the user has any permission for.
+            self._pages = (
+                Page.objects.filter(
+                    group_permissions__group__user=user,
+                )
+                .prefetch_related(prefetch_user_perms)
+                .distinct()
+                .in_bulk()
+            )
+            # Since a lot of queries involve looking up pages
+            # the user has a specific permission for, we
+            # build this mapping of `permission_type: {page_pks}`
+            # to retrieve that information quickly without asking
+            # the database.
+            self.perm_types = defaultdict(set)
+            for page_pk, page in self._pages.items():
+                # Rewrite the `_perms_for_user` annotation and only
+                # store the different type of permission the user has.
+                page._perms_for_user = {
+                    perm.permission_type for perm in page._perms_for_user
+                }
+
+                for perm_type in page._perms_for_user:
+                    self.perm_types[perm_type].add(page_pk)
+
+        return user._page_permissions_proxy
+
+    def get_pages_for_perms(self, perm_types):
+        """
+        Returns an iterator over all the pages the user has permissions matching `perm_types`.
+        """
+        page_pks = set()
+        for perm_type in perm_types:
+            pages_for_perm = self.perm_types.get(perm_type, None)
+            if pages_for_perm is not None:
+                page_pks |= pages_for_perm
+
+        return (self._pages[page_pk] for page_pk in page_pks)
+
+    @property
+    def pages(self):
+        """
+        Returns an iterator over all the pages the user has any permission for.
+        """
+        return (page for page in self._pages.values())
+
+    def has_any_page_permission(self):
+        return bool(self._pages)
 
-        if user.is_active and not user.is_superuser:
-            self.permissions = GroupPagePermission.objects.filter(
-                group__user=self.user
-            ).select_related("page")
+    def for_page(self, page):
+        """
+        Return a PagePermissionTester object that can be used to query whether this user has
+        permission to perform specific tasks on the given page.
+        """
+        return PagePermissionTester(self, page)
+
+    def pages_with_direct_explore_permission(self):
+        # Get all pages that the user has direct add/edit/publish/lock permission on
+        if self.user.is_superuser:
+            # superuser has implicit permission on the root node
+            return Page.objects.filter(depth=1)
+        else:
+            return self.get_pages_for_perms(["add", "edit", "publish", "lock"])
+
+    @functools.lru_cache(maxsize=None)
```

Comment:
Suggestion: Rather than using `lru_cache`, it might be better to manually cache this in an attribute. LRU caches can hang around longer than you'd expect when attached to classes, and could cause cache invalidation issues. It should also hopefully avoid any memory bloat issues due to `maxsize=None`.

Reply:
That makes sense. I've applied your suggestion, thanks!